File: services/receipt-processor/app/models/receipts.py
import json
from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field, field_validator


class ProcessingStatus(StrEnum):
    """Supported receipt processing statuses."""

    uploaded = "uploaded"
    processing = "processing"
    completed = "completed"
    failed = "failed"


class ReceiptCreate(BaseModel):
    """Internal receipt creation payload."""

    id: str
    user_id: str
    original_filename: str
    storage_path: str
    mime_type: str
    file_size: int


class ReceiptItemCreate(BaseModel):
    """Line item extracted from a receipt."""

    receipt_id: str
    user_id: str
    description: str
    quantity: Decimal | None = None
    unit_price: Decimal | None = None
    total: Decimal | None = None


class ReceiptRecord(BaseModel):
    """Stored receipt record."""

    id: str
    user_id: str
    merchant: str = "Processing receipt"
    receipt_date: date
    category: str = "Uncategorised"
    total: Decimal = Decimal("0")
    currency: str = "GBP"
    confidence: int = Field(default=0, ge=0, le=100)
    status: str = "processing"
    is_business: bool = False
    original_filename: str
    storage_path: str
    mime_type: str
    file_size: int
    processing_status: ProcessingStatus
    raw_ocr_text: str | None = None
    merchant_name: str | None = None
    transaction_date: date | None = None
    subtotal: Decimal | None = None
    tax: Decimal | None = None
    ai_summary: str | None = None
    embedding: list[float] | None = None
    ocr_provider: str | None = None
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    @field_validator("embedding", mode="before")
    @classmethod
    def parse_pgvector(cls, value: object) -> object:
        """Parse PostgREST pgvector text output into a float list."""
        if value is None or isinstance(value, list):
            return value
        if isinstance(value, str):
            parsed = json.loads(value)
            if isinstance(parsed, list):
                return parsed
        return value


class ReceiptResponse(BaseModel):
    """Public receipt response."""

    receipt: ReceiptRecord


class HealthResponse(BaseModel):
    """Health check response."""

    status: str = "ok"
    version: str


class FileValidationResult(BaseModel):
    """Validated upload metadata and bytes."""

    filename: str
    mime_type: str
    content: bytes = Field(repr=False)

    @property
    def size(self) -> int:
        return len(self.content)

File: lib/receipts.ts
import "server-only";

import { createClient } from "@/lib/supabase/server";
import { mockDashboardData } from "@/lib/mock-data";
import type { ReceiptStatus } from "@/lib/types";

const configured = Boolean(
  process.env.NEXT_PUBLIC_SUPABASE_URL &&
    process.env.NEXT_PUBLIC_SUPABASE_PUBLISHABLE_KEY
);

export type ReceiptListItem = {
  id: string;
  merchant: string;
  receipt_date: string;
  category: string;
  total: number | string;
  currency: string;
  confidence: number;
  status: ReceiptStatus;
  is_business: boolean;
  processing_status: "uploaded" | "processing" | "completed" | "failed" | null;
  original_filename: string | null;
  created_at: string;
  updated_at: string | null;
};

export type ReceiptDetail = ReceiptListItem & {
  storage_path: string | null;
  mime_type: string | null;
  file_size: number | null;
  raw_ocr_text: string | null;
  merchant_name: string | null;
  transaction_date: string | null;
  subtotal: number | string | null;
  tax: number | string | null;
  ai_summary: string | null;
  ocr_provider: string | null;
  error_message: string | null;
  items: ReceiptItem[];
};

export type ReceiptItem = {
  id: string;
  description: string;
  quantity: number | string | null;
  unit_price: number | string | null;
  total: number | string | null;
};

export type ReceiptReasoningRecord = ReceiptListItem & {
  items: ReceiptItem[];
};

const listSelect = `
  id,
  merchant,
  receipt_date,
  category,
  total,
  currency,
  confidence,
  status,
  is_business,
  processing_status,
  original_filename,
  created_at,
  updated_at
`;

const detailSelect = `
  ${listSelect},
  storage_path,
  mime_type,
  file_size,
  raw_ocr_text,
  merchant_name,
  transaction_date,
  subtotal,
  tax,
  ai_summary,
  ocr_provider,
  error_message
`;

export async function getReceiptList(): Promise<ReceiptListItem[]> {
  if (!configured) {
    return mockDashboardData.recent.map((receipt) => ({
      ...receipt,
      currency: "GBP",
      processing_status:
        receipt.status === "review" ? null : receipt.status,
      original_filename: null,
      updated_at: null,
    }));
  }

  const supabase = await createClient();

  const { data, error } = await supabase
    .from("receipts")
    .select(listSelect)
    .order("created_at", { ascending: false });

  if (error) {
    throw new Error(`Unable to load receipts: ${error.message}`);
  }

  return (data ?? []) as ReceiptListItem[];
}

export async function getReceiptDetail(
  id: string
): Promise<ReceiptDetail | null> {
  if (!configured) {
    const receipt = mockDashboardData.recent.find((item) => item.id === id);

    if (!receipt) {
      return null;
    }

    return {
      ...receipt,
      currency: "GBP",
      processing_status:
        receipt.status === "review" ? null : receipt.status,
      original_filename: null,
      updated_at: null,
      storage_path: null,
      mime_type: null,
      file_size: null,
      raw_ocr_text: null,
      merchant_name: null,
      transaction_date: null,
      subtotal: null,
      tax: null,
      ai_summary: null,
      ocr_provider: null,
      error_message: null,
      items: [],
    };
  }

  const supabase = await createClient();

  const { data, error } = await supabase
    .from("receipts")
    .select(detailSelect)
    .eq("id", id)
    .single();

  if (error) {
    if (error.code === "PGRST116") {
      return null;
    }

    throw new Error(`Unable to load receipt: ${error.message}`);
  }

  const { data: items, error: itemsError } = await supabase
    .from("receipt_items")
    .select("id, description, quantity, unit_price, total")
    .eq("receipt_id", id)
    .order("created_at", { ascending: true });

  if (itemsError) {
    throw new Error(`Unable to load receipt items: ${itemsError.message}`);
  }

  return {
    ...(data as Omit<ReceiptDetail, "items">),
    items: (items ?? []) as ReceiptItem[],
  };
}

export async function getReceiptReasoningRecords(
  limit = 25
): Promise<ReceiptReasoningRecord[]> {
  const receipts = (await getReceiptList()).slice(0, limit);

  if (!configured || receipts.length === 0) {
    return receipts.map((receipt) => ({
      ...receipt,
      items: [],
    }));
  }

  const supabase = await createClient();
  const receiptIds = receipts.map((receipt) => receipt.id);

  const { data: items, error } = await supabase
    .from("receipt_items")
    .select("id, receipt_id, description, quantity, unit_price, total")
    .in("receipt_id", receiptIds)
    .order("created_at", { ascending: true });

  if (error) {
    throw new Error(`Unable to load receipt reasoning items: ${error.message}`);
  }

  const itemsByReceipt = new Map<string, ReceiptItem[]>();

  (items ?? []).forEach((item) => {
    const receiptId = String(item.receipt_id);
    const currentItems = itemsByReceipt.get(receiptId) ?? [];
    currentItems.push({
      id: item.id,
      description: item.description,
      quantity: item.quantity,
      unit_price: item.unit_price,
      total: item.total,
    });
    itemsByReceipt.set(receiptId, currentItems);
  });

  return receipts.map((receipt) => ({
    ...receipt,
    items: itemsByReceipt.get(receipt.id) ?? [],
  }));
}
